aggregator/map.go

```diff
@@ -0,0 +1,193 @@
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package aggregator
+
+import (
+	"container/list"
+	"math"
+	"sync"
+	"time"
+
+	"github.com/m3db/m3aggregator/id"
+	"github.com/m3db/m3metrics/metric"
+	"github.com/m3db/m3metrics/metric/unaggregated"
+	"github.com/m3db/m3metrics/policy"
+	"github.com/m3db/m3x/clock"
+)
+
+var (
+	emptyHashedEntry hashedEntry
+)
+
+type hashedEntry struct {
+	idHash id.Hash
+	entry  *Entry
+}
+
+// NB(xichen): use a type-specific list for hashedEntry if the conversion
+// overhead between interface{} and hashedEntry becomes a problem
+type metricMap struct {
+	sync.RWMutex
+
+	opts         Options
+	nowFn        clock.NowFn
+	entryPool    EntryPool
+	batchPercent float64
+	doneCh       <-chan struct{}
+
+	metricLists *MetricLists
+	entries     map[id.Hash]*list.Element
+	entryList   *list.List
+	waitForFn   waitForFn
+}
+
+func newMetricMap(lists *MetricLists, doneCh <-chan struct{}, opts Options) *metricMap {
+	return &metricMap{
+		opts:         opts,
+		nowFn:        opts.ClockOptions().NowFn(),
+		entryPool:    opts.EntryPool(),
+		batchPercent: opts.EntryCheckBatchPercent(),
+		doneCh:       doneCh,
+		metricLists:  lists,
+		entries:      make(map[id.Hash]*list.Element),
+		entryList:    list.New(),
+		waitForFn:    time.After,
+	}
+}
+
+func (m *metricMap) AddMetricWithPolicies(
+	mu unaggregated.MetricUnion,
+	policies policy.VersionedPolicies,
+) error {
+	e := m.findOrCreate(mu.ID)
+	err := e.AddMetricWithPolicies(mu, policies)
+	e.DecWriter()
+	return err
+}
+
+func (m *metricMap) DeleteExpired(target time.Duration) {
+	now := m.nowFn()
+
+	// Determine batch size
+	m.RLock()
+	elemsLen := m.entryList.Len()
+	batchSize := int(math.Max(1.0, math.Ceil(m.batchPercent*float64(elemsLen))))
+	numBatches := int(math.Ceil(float64(elemsLen) / float64(batchSize)))
+	var targetPerBatch time.Duration
+	if numBatches > 0 {
+		targetPerBatch = target / time.Duration(numBatches)
+	}
+	currElem := m.entryList.Front()
+	m.RUnlock()
+
+	// NB(xichen): if this runs frequently enough, stash the expired buffer
+	// in the map object and reuse the buffer
+	expired := make([]hashedEntry, 0, batchSize)
+	batchIdx := 0
+	for currElem != nil {
+		m.RLock()
+		for numChecked := 0; numChecked < batchSize && currElem != nil; numChecked++ {
+			nextElem := currElem.Next()
```

Comment:
elements can only be added to the end of the list after acquiring a write lock. As long as the batch size is positive and close enough to the optimal batch size, we'll eventually get through the list.